### PR TITLE
feat: add parameter to for output stream

### PR DESCRIPTION
**Pull request by bpshaver**
The addition of the `stream` parameter allows the user to choose to send the yaspin spinner output to a stream or TextIO object other than sys.stdout. The most obvious use for this feature is to set the stream to stderr to support piping of output like `python main.py > some.log` and seeing the yaspin spinner in the terminal while stdout goes to `some.log`

### CODE REVIEW

File: yaspin/core.py [hunk 1]
```diff
@@ -591,24 +612,3 @@ def _set_interval(spinner: Spinner) -> float:
     def _set_cycle(frames: Union[str, Sequence[str]]) -> Iterator[str]:
         return itertools.cycle(frames)
 
-    @staticmethod
```
Comment (bpshaver) [inline yaspin/core.py:594]: Because of calls so `isatty` these can no longer be static methods

**Comment by pavdmyt** [comment]
@bpshaver please run `make fmt` on your changes so that `check-fmt` check can succeed in CI.

**Comment by bpshaver** [comment]
> @bpshaver please run `make fmt` on your changes so that `check-fmt` check can succeed in CI.

Done.

Does this seem like a sensible change? If so, I'll look through the failing test cases now.

**Comment by pavdmyt** [comment]
@bpshaver thanks for your contribution! I'll try to review this thoroughly next week.

**Comment by pavdmyt** [comment]
Apparently it is possible to detect situations when different streams are not tty:
```
import os

if not os.isatty(sys.stdout.fileno()):
    print("The output stream is not a TTY.")
else:
    print("The output stream is a TTY.")
```
So, I would rather prefer this approach, aligning behavior with e.g. https://github.com/pavdmyt/yaspin/issues/31#issuecomment-1016861612 to move further here. It must be much more convenient than requiring developer to define custom streams upfront. Plus, if yaspin is capable to find out which streams it is attached to, there is no need to rewrite older code.

**Comment by bpshaver** [comment]
@pavdmyt I do not understand how this is more convenient. The idea of detecting whether the target stream is a TTY and changing behavior or issuing a warning is a good one. But whether that stream is _always_ stdout or whether it should be configurable by the user is a separate question. 

If anything, https://github.com/pavdmyt/yaspin/issues/31#issuecomment-1016861612 supports making `sys.stderr` the default stream, _not_ `sys.stdout`. I am suggesting a less drastic change, which is simply that `yaspin` open up the possibility for a user to send the spinner to stderr. And I disagree with your comment:

> It must be much more convenient than requiring developer to define custom streams upfront.

In fact, my proposed change keeps `sys.stdout` as the default value (when `stream` is `None`)  so it does not require anything from a developer who wants to continue using `yaspin` as before. The change is completely backwards compatible. And there is no need to define custom streams. stdout or stderr should both be compatible streams and neither are "custom."

**Comment by pavdmyt** [comment]
I think, I need to elaborate a bit more about my point. The idea behind capability to automatically detect streams to which yaspin is attached to allows to separate:
 - _spinner's stream_ (escape and color codes, spinner frames) from the 
 - _useful text_ (`.text` attribute, `write()` and possibly `.ok()` and `.fail()` methods)

So that _spinner's stream_ is always ends up in `stderr` (backwards incompatible change) and the _useful text_ in `stdout`. This way, you can make decisions which stream ends up where by simply using redirects and the default behavior stays (visually) the same, as both `stderr` and `stdout` are attached to the TTY by default. This would allow doing `python foo.py > foo.log` (from your example above) having the _useful text_ inside `foo.log`, while _spinner's stream_ got rendered in the terminal window (as `stderr` remained attached to TTY). And this doesn't require any code changes using `stream` argument or attribute.

To summarize, we want to have _spinner's stream_ and _useful text_ clearly separated so that user can decide what to do with these streams. Having such feature, I'm not sure about the value of having an additional `stream` argument (please correct me if I'm wrong or in case I'm missing something).

**Comment by bpshaver** [comment]
I understand now, thanks!

I would call the `.text`, etc. _accompanying text_. Under normal circumstances, it is right next to the spinner. You're assuming that when the output is redirected to a non-TTY stream, however, the _accompanying text_ should be interpreted as _useful text_ and go to that non-TTY stream. But I'm not so sure that assumption makes sense for all users.

In the issue I was trying to tackle ( https://github.com/mandiant/capa/issues/1812 ) the accompanying text should go to `stderr` (remaining next to the spinner graphic) while useful text going to the non-TTY stream is everything contained within the spinner context:

```bash
$ python foo.py > foo.log
⠋ Analyzing data...
```

while program output goes to `foo.log`.

I understand a sensible default may be to send accompanying text to the non-TTY stream, but this example I think illustrates a case where that is not desired.

**Comment by pavdmyt** [comment]
> In the issue I was trying to tackle ( https://github.com/mandiant/capa/issues/1812 ) the accompanying text should go to stderr (remaining next to the spinner graphic) while useful text going to the non-TTY stream is everything contained within the spinner context

I'm not sure I've understood this correctly, "accompanying text should go to `stderr` while useful text going to the non-TTY stream" - isn't accompanying and useful text are the same entities? It is also possible to redirect `stderr` to the non-TTY stream.

**Comment by bpshaver** [comment]
_accompanying text_ is my name for whatever text normally goes right next to the spinner. You called it _useful text_, which is begging the question with respect to where it should go in the corner cases we're discussing. Call them what you like, but I'm pointing to three different types of text (not two):

- The actual spinner - should probably never go to a non-TTY stream
- Some text to the right of the spinner (`.text`, etc.) - maybe should go to a non-TTY stream. Probably depends on the user.
- Normal text written as result of `print` calls and the like - should be redirected as normal, unaffected by being within a `yaspin` spinner context.

File: yaspin/core.py [hunk 2]
```diff
@@ -374,15 +378,11 @@ def fail(self, text: str = "FAIL") -> None:
         _text = text if text else "FAIL"
         self._freeze(_text)
 
+    def is_jupyter(self) -> bool:
```
Comment (odisseus) [inline yaspin/core.py:381]: I am pretty sure that Jupyter is not the only front-end capable of intercepting the output stream.

File: yaspin/core.py [hunk 3]
```diff
@@ -533,6 +544,15 @@ def _set_attrs(attrs: Sequence[str]) -> set[str]:
                 )
         return set(attrs)
 
+    # Static
```
Comment (odisseus) [inline yaspin/core.py:547]: What is the purpose of this comment?

**Comment by odisseus** [comment]
What is the status of this pull request? How can I help getting it merged?

**Comment by bpshaver** [comment]
Seems like @pavdmyt disagreed with my reasoning and the discussion ended.

**Comment by darkpanda0** [comment]
We are in the process of migrating from the [halo](https://pypi.org/project/halo/) spinner, because it is no longer actively maintained. yaspin was the obvious choice, but I'm not sure if we will be able to use it without this feature. Our spinner is shared across various components that  individual developers can take advantage of, if the control of the flow passes to their module. I have a wrapper class to minimize the code changes necessary for developers. Some of the time our app is instructed to output json. When I am redirecting to a file, I am still getting spinner messages in the json file. [halo](https://pypi.org/project/halo/) simply allows us to choose sys.stderr, so not having that option is making it look like the transition to yaspin might not work.

I'm currently testing with python 3.11.4 using yaspin 3.1.0

**Comment by bpshaver** [comment]
@darkpanda0 I share your concerns. I just updated my PR with the latest from upstream so that @pavdmyt can re-review. Maybe you can try my branch to see if it works for you?

Can we get this merged in? @odisseus would probably like to see it merged as well.

File: tests/test_in_out.py [hunk 4]
```diff
@@ -120,7 +120,13 @@ def test_hide_show(monkeypatch, capsys, text, request, isatty_fixture):
 
     # Ensure that sp.stop() will be executed
     def teardown():
-        sp.stop()
+        try:
+            sp.stop()
```
[the remark this answers is not in the document]
Comment (pavdmyt) [reply tests/test_in_out.py:128]: The issue is that `capsys` closes the file handles during teardown, but the spinner is still trying to write to them.

E.g. during test teardown, `capsys` closes the `stdout` stream, but when `sp.stop()` is called, it tries to call `_clear_line()` which attempts to write to the closed stream.

**Review by pavdmyt — CHANGES_REQUESTED** [review]
We cannot simply suppress errors and ignore test failures, needs to be fixed.

File: yaspin/core.py [hunk 5]
```diff
@@ -645,15 +646,28 @@ def _reset_signal_handlers(self) -> None:
...
+        if self._stream.isatty():
+            # ANSI Control Sequence EL does not work in Jupyter
+            self._stream.write("\r\033[K")
+        else:
+            fill = " " * self._cur_line_len
+            self._stream.write(f"\r{fill}\r")
+
+    def _set_color(self, value: str) -> str:
+        if self.is_jupyter():
             Yaspin._warn_color_disabled()
```
[the remark this answers is not in the document]
Comment (pavdmyt) [reply yaspin/core.py:686]: `self._stream` call is the first thing in constructor, this is also well-tested.
